### ISSUE TITLE
S1128 improve performance

### ISSUE THREAD

**Issue by andrei-epure-sonarsource** [SonarSource/sonar-dotnet#3761]
S1128 (UnnecessaryUsings) is a code smell which doesn't bring a lot of functional value, so we don't want it to have a big performance impact on the msbuild time.

We should re-think how we use the Semantic Model inside the rule, in order to improve the performance.

From this [investigation](https://community.sonarsource.com/t/re-net-analysis-time-during-msbuild-increased-from-15-min-to-40-min/32956/3), the rule takes 38 seconds to run - 4%.

> 38.864    4      SonarAnalyzer.Rules.CSharp.UnnecessaryUsings

**Comment by Corniel** [comment]
Improving performance is a good thing! Personally I value this rule highly, as it potentially removes a lot of code, which makes code easier to read. Visual Studio also has tooling for this, I'm not sure if sonar can build on top of that (might help performance wise).

**Comment by andrei-epure-sonarsource** [comment]
I agree that performance is important, @Corniel. As our analyzers are used on CI systems, as well as local builds outside the IDE - we cannot rely on Visual Studio. It would be interesting to see their implementation, though.

**Comment by pavel-mikula-sonarsource** [comment]
This rule needs complete redesign. New idea:
* Walk the syntax tree
* Collect imported symbol from `using` statement. 
* Track all imported members from collected symbol (like `FileStream` from `using System.IO;`)
* Do not ask for symbol of tracked identifiers

**Comment by mary-georgiou-sonarsource** [comment]
Leaving this here as an FYI:
Global using retrieval will be available from MSBuild version 17.3 onwards through [this ](https://github.com/CyrusNajmabadi/roslyn/blob/main/src/Compilers/Core/Portable/Compilation/IImportScope.cs) API,  exposed via `semanticModel.GetImportScopes`.

**Comment by pavel-mikula-sonarsource** [comment]
Moved to [NET-1939](https://sonarsource.atlassian.net/browse/NET-1939)

[NET-1939]: https://sonarsource.atlassian.net/browse/NET-1939?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
